docs/programmers_guide/witness_formal_spec.md

```diff
@@ -106,63 +201,103 @@ Helper functions MUST have at least one argument.
+```
 
-Our example witness would be `HASH h1; HASH h2; BRANCH 0b11`.
+### `EXTENSION key`
 
-Initial state of the stack is ` <empty> `;
+**Substitution rules**
 
----
+```
+GUARD node != nil
 
-**1. Executing `HASH h1`**: push a `hashNode` to the stack.
+Node(node) EXTENSION(key) |=> ExtensionNode{key, node}
 
-Witness: `HASH h2; BRANCH 0b11`
+```
 
-Stack: `(hashNode(h1), h1)`
+### `HASH raw_hash`
 
----
+Pushes a `HashNode` to stack.
 
-**2. Executing `HASH h2`**: push one more `hashNode` to the stack.
+**Substitution rules**
 
-Witness `BRANCH 0b11`
+```
+HASH(hash_value) |=> HashNode{hash_value}
+```
+
+### `CODE raw_code`
+
+Pushes an nil node + the code hash to the stack.
+
+```
+CODE(raw_code) |=> CodeNode{raw_code}
+```
 
-Stack: `(hashNode(h2), h2); (hashNode(h1), h1)`
+### `ACCOUNT_LEAF key nonce balance has_code has_storage`
+
+**Substitution rules**
+
+```
+GUARD has_code == true
+GUARD has_storage == true
+
+CodeNode(code) Node(storage_hash_node) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, storage_root, code}}
 
 ---
 
-**3. Executing `BRANCH 0b11`**: replace 2 items on the stack with a single `fullNode`.
+GUARD has_code == true
+GUARD has_storage == true
+
+HashNode(code) Node(storage_hash_node) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, storage_root, code}}
+
+---
 
-Witness: ` <empty> `
+GUARD has_code == false
+GUARD has_storage == true
 
-Stack: `(fullNode{0: hashNode(h2), 1: hashNode(h1)}, KECCAK(h2+h1))`
+Node(storage_root) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, storage_root, nil}}
 
 ---
 
-So our stack has exactly a single item and there are no more instructions in the witness, the execution is completed.
+GUARD has_code == true
+GUARD has_storage == false
 
-## Modes
+CodeNode(code) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, nil, nil, code}}
 
-There are two modes of execution for this stack machine:
+---
 
-(1) **normal execution** -- the mode that constructs a trie;
+GUARD has_code == true
+GUARD has_storage == false
 
-(2) **hash only execution** -- the mode that calculates the root hashes of a trie without constructing the trie itself;
+HashNode(code) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, nil, nil, code}}
 
-In the mode (2), a stack item  MUST only contain the node hash.
+---
 
-Stack item, mode (1): `(node, hash)`.
+GUARD has_code == false
+GUARD has_storage == false
 
-Stack item, mode (2): `(hash)`.
+ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, nil, nil, nil}}
 
-The exact implementation details are undefined in this spec.
+```
 
-## Instructions
+### `NEW_TRIE`
+
+Stops the witness execution.
```

Comment:
It does not stop the execution, it is just something that does not get substituted, because there is no rules matching it